### PR TITLE
Added support for cluster level failouts (SLF)

### PR DESCRIPTION
**Pull request by cstufflebeam**

### CODE REVIEW

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java
Comment (jpstewart) [inline d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java:120]: Style nit, I believe the coding style for Pegasus prefers arguments on the same line with breaks only to stay under the line limit.

**Review by jpstewart — APPROVED** [review]
Looks good to me, just style nits. Be sure to get approval from someone more familiar with the area like Bohan or Rick though.

**Review by bohhyang — CHANGES_REQUESTED** [review]
Since the ClusterStoreProperties will include failout properties, please also add a test in ClusterLoadBalancerSubscriberTest to verify it will work with the new data.

**Comment by bohhyang** [comment]
Noticed that some tests are failing. Please run them locally a few times, and if they are just being flaky, please add 3 retries referencing: https://iwww.corp.linkedin.com/wiki/cf/pages/viewpage.action?spaceKey=ENGS&title=Pegasus+Git+Workflow#PegasusGitWorkflow-DealingwithFlakyCITests

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterFailoutProperties.java
Comment (bbarkley) [inline d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterFailoutProperties.java:32]: It'd be much better to have these as strongly typed objects instead of Maps (e.g. List<FailoutRedirectConfig> and List<FailoutBucketConfig>). You'll need to write more code in the serializer/deserializer, but the type safety and self documenting code is worth it. Or is this being done to keep LI specific properties out of the OS code? If that's the case where will these be consumed, and will there be a type safe Java object representing them?

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java
Comment (cstufflebeam) [reply d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java:223]: Fixed, thanks.

File: d2/src/test/java/com/linkedin/d2/balancer/properties/ClusterPropertiesSerializerTest.java
Comment (cstufflebeam) [reply d2/src/test/java/com/linkedin/d2/balancer/properties/ClusterPropertiesSerializerTest.java:248]: Added.

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterFailoutProperties.java
Comment (cstufflebeam) [reply d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterFailoutProperties.java:32]: Hey Bryan - we're intentionally keeping this weakly typed, in order to avoid exposing LI concepts in rest.li. The strongly typed properties are provided in a class in a library that we use internally.

**Review by bohhyang — APPROVED** [review]
thanks for addressing the comments. Pls also wait for Bryan and Rick's approval . Thanks.

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterFailoutProperties.java
Comment (bbarkley) [reply d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterFailoutProperties.java]: Right, forgot about that decision. Thanks for the reminder.

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java
Comment (bbarkley) [reply d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java:134]: The style guide is to minimize the line breaks, so have as many arguments on each line as possible, and break when it goes over the limit as John indicated.

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java
Comment (bbarkley) [inline d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java:230]: Our general Java style guide requires braces around control statements for single lines.

**Review by bbarkley — CHANGES_REQUESTED** [review]
Minor style issues, but otherwise looks good to me.

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java
Comment (cstufflebeam) [reply d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java:134]: All of the other preexisting constructors in this file are using one line per argument; Should I go ahead and change all of them?

Comment (bbarkley) [reply d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java:134]: If that's how the rest of the file is I think it's fine to leave them as-is

**Review by bbarkley — APPROVED** [review]

File: d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterPropertiesJsonSerializer.java
Comment (rickzx) [inline d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterPropertiesJsonSerializer.java:100]: `FailoutProperties` -> `failoutProperties`

**Review by rickzx — COMMENTED** [review]
Thanks, looks good. One minor comment.

Let's also update the CHANGELOG and version. Please refer to: https://iwww.corp.linkedin.com/wiki/cf/pages/viewpage.action?spaceKey=ENGS&title=Pegasus+Git+Workflow#PegasusGitWorkflow-Releasing

**Review by rickzx — APPROVED** [review]

**Pull request merged**
